# TP NoSQL

Prénom :MAXIMILIEN

Nom : SOUMAHORO

Pour ce TP noté vous avez trois heures pour le réaliser et 48 heures pour le finir si besoin. Le rendu final est attendu pour dimanche 29 septembre 2024 à 23 h 59.

Vous serrez évalués sur les requêtes mais également sur la qualité et la propreté du code. Pour chaque question, une réponse présentant les résultats est attendue.

In [1]:
# installation des packages pandas et pymongo
!pip install -r requirements.txt --quiet

ERROR: Could not install packages due to an OSError: [WinError 5] Accès refusé: 'C:\\Anaconda\\Lib\\site-packages\\~andas.libs\\msvcp140-ef6047a69b174ada5cb2eff1d2bc9a62.dll'
Consider using the `--user` option or check the permissions.



In [2]:
import pymongo
import pandas as pd

URI = 'mongodb+srv://mongo_user:Vfd939J3hPgAUNox@cluster-but-sd.tpg0l.mongodb.net/?retryWrites=true&w=majority&appName=cluster-but-sd'
client = pymongo.MongoClient(URI)
db = client.tp

# output the name of the collections in the database
print("Collections: ", db.list_collection_names())

C:\Anaconda\Lib\site-packages\pandas\core\arrays\masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (


Collections:  ['restaurants', 'movies']


1. Combien de films sont dans la base de données ?

In [13]:
db.movies.count_documents({})

23530

Nous avons 23530 films dans la base de données


2. Quel est le nombre moyen de votes par pays ? Trier les pays par nombre moyen de votes décroissant.

In [23]:
a= db.movies.aggregate([
    {"$unwind": "$countries"},  
    {"$group": {
        "_id": "$countries",  
        "Moyenne": {"$avg": "$imdb.votes"}  
    }},
    {"$sort": {"MoyenneDesVotes": -1}}  
])
pd.DataFrame(a)

_id       Moyenne
0            Macao   3949.000000
1           Russia   3106.769231
2    Faroe Islands    436.000000
3             Peru  13401.764706
4            China  15653.136667
..             ...           ...
153       Colombia   1647.393939
154         Panama   4791.500000
155        Nigeria    711.500000
156         Bhutan   2115.000000
157    El Salvador    675.000000

[158 rows x 2 columns]

3. Donner la liste des différents genres de films.

In [24]:
genres = db.movies.aggregate([
    {"$unwind": "$genres"}, 
    {"$group": {
        "_id": "$genres"  
    }},
    {"$sort": {"_id": 1}} 
])


pd.DataFrame(genres)

_id
0        Action
1     Adventure
2     Animation
3     Biography
4        Comedy
5         Crime
6   Documentary
7         Drama
8        Family
9       Fantasy
10    Film-Noir
11      History
12       Horror
13        Music
14      Musical
15      Mystery
16         News
17      Romance
18       Sci-Fi
19        Short
20        Sport
21    Talk-Show
22     Thriller
23          War
24      Western

4. Combien de films ont une note IMDb supérieure à 8 ?

In [25]:
IMDb8 = db.movies.count_documents({"imdb.rating": {"$gt": 8}})
IMDb8

1177

Le nombre de  films ayant une note IMDb supérieure à 8 est de 1177

5. Afficher le titre, l'année de sortie et les réalisateurs de tous les films (ne pas afficher l'ID).

In [29]:
a = db.movies.find(
    {},
    {"_id": 0, "title": 1, "year": 1, "directors": 1}
)

pd.DataFrame(a)


title  \
0      Intolerance: Love's Struggle Throughout the Ages   
1                             The Poor Little Rich Girl   
2                                               The Kid   
3                                        Miss Lulu Bett   
4                                          Safety Last!   
...                                                 ...   
23525                                            Emelie   
23526                                    Techo y comida   
23527                                             Nahid   
23528                                  Coin Locker Girl   
23529                                 Tryapichnyy soyuz   

                            directors  year  
0                     [D.W. Griffith]  1916  
1                  [Maurice Tourneur]  1917  
2                   [Charles Chaplin]  1921  
3               [William C. de Mille]  1921  
4      [Fred C. Newmeyer, Sam Taylor]  1923  
...                               ...   ...  
23525                [Michael Thelin]  2015  
23526      [Juan Miguel del Castillo]  2015  
23527                [Ida Panahandeh]  2015  
23528                   [Jun-hee Han]  2015  
23529            [Mikhail Mestetskiy]  2015  

[23530 rows x 3 columns]

6. Afficher le titre et le plot des films réalisés par D.W. Griffith.

In [30]:
a = db.movies.find(
    {"directors": "D.W. Griffith"},
    {"_id": 0, "title": 1, "plot": 1}
)

pd.DataFrame(a)

plot  \
0  The story of a poor young woman, separated by ...   
1  A greedy tycoon decides, on a whim, to corner ...   
2  A frail waif, abused by her brutal boxer fathe...   
3  The Civil War divides friends and destroys fam...   

                                              title  
0  Intolerance: Love's Struggle Throughout the Ages  
1                                 A Corner in Wheat  
2    Broken Blossoms or The Yellow Man and the Girl  
3                             The Birth of a Nation

7. Quels sont les 5 acteurs les plus populaires (ceux qui ont joué dans le plus de films) ?

In [34]:
acteur_populaire = db.movies.aggregate([
    {"$unwind": "$cast"},
    {"$group": {
        "_id": "$cast",
        "count": {"$sum": 1}
    }},
    {"$sort": {"count": -1}},
    {"$limit": 5}
])

pd.DataFrame(acteur_populaire)

_id  count
0      Gèrard Depardieu     68
1        Robert De Niro     60
2         Michael Caine     52
3  Marcello Mastroianni     50
4         Max von Sydow     49

8. Quelles sont les 6 années avec le plus de films produits ?

In [35]:
six_meilleurs = db.movies.aggregate([
    {"$group": {
        "_id": "$year",
        "count": {"$sum": 1}
    }},
    {"$sort": {"count": -1}},
    {"$limit": 6}
])

pd.DataFrame(six_meilleurs)

_id  count
0  2013   1220
1  2014   1147
2  2012   1109
3  2011   1040
4  2009    990
5  2010    970

9. Combien de films ont été réalisés avant l'année 2000 et ont une note IMDb supérieure à 7 ?

In [36]:
a = db.movies.count_documents({
    "year": {"$lt": 2000},
    "imdb.rating": {"$gt": 7}
})
a


4747

10. Donner les 20 réalisateurs qui ont réalisé le plus de films (afficher seulement leur nom et le nombre de films).

In [37]:
a = db.movies.aggregate([
    {"$unwind": "$directors"},
    {"$group": {
        "_id": "$directors",
        "count": {"$sum": 1}
    }},
    {"$sort": {"count": -1}},
    {"$limit": 20}
])

pd.DataFrame(a)


_id  count
0         Woody Allen     40
1           John Ford     35
2       Takashi Miike     34
3         John Huston     34
4       Werner Herzog     33
5     Martin Scorsese     32
6    Alfred Hitchcock     31
7        Sidney Lumet     30
8    Steven Spielberg     29
9     Mario Monicelli     29
10      Michael Apted     29
11       George Cukor     29
12  Steven Soderbergh     28
13          Spike Lee     28
14      Robert Altman     28
15     Clint Eastwood     27
16    Jean-Luc Godard     27
17        Wim Wenders     27
18          Ken Loach     27
19         Johnnie To     27

11. Quelle est la note IMDb moyenne des films sortis après 2010 ?

In [44]:
moy_2010 = db.movies.aggregate([
    {"$match": {"year": {"$gt": 2010}}},
    {"$group": {
        "_id": None,
        "average_imdb_rating": {"$avg": "$imdb.rating"}
    }}
])
moy_2010





6.544033171521035

12. (bonus) Afficher les films dont le titre contient le mot "love".

In [46]:
love = db.movies.find(
    {"title": {"$regex": "love", "$options": "i"}},
    {"_id": 0, "title": 1}
)

pd.DataFrame(love)


title
0    Intolerance: Love's Struggle Throughout the Ages
1                                      Arise, My Love
2                                   One Night of Love
3                               Love Finds Andy Hardy
4                                     Love Me Tonight
..                                                ...
438                                        Much Loved
439                               Cafè. Waiting. Love
440                    Larry Kramer in Love and Anger
441                                    Our Loved Ones
442                   My Love, Don't Cross That River

[443 rows x 1 columns]

13. (bonus) Afficher le titre des films ayant un score Rotten Tomatoes (viewer) supérieur à 75.


In [57]:
reponse13 = db.movies.aggregate([
    {
        "$match": {"tomatoes.viewer.meter": {"$gt": 75}}  
    },
    {
        "$project": {
            "_id": 0, 
            "title": 1 
        }
    }
])

pd.DataFrame(list(reponse13))

title
0     Intolerance: Love's Struggle Throughout the Ages
1                            The Poor Little Rich Girl
2                                              The Kid
3                                         Safety Last!
4                                         Wild Oranges
...                                                ...
6597                                             Luton
6598                                       Amira & Sam
6599                                     Misconception
6600                                             Grace
6601                         Batman: Assault on Arkham

[6602 rows x 1 columns]